lib/generator.js

```diff
@@ -7,15 +7,140 @@
 
 /**
  * @typedef {object} Configuration
- * @property {boolean} overwrite - will overwrite the target directory if true, by default false
- * @property {object} variables - variables to be resolved in ejs template
+ * @property {boolean} [overwrite=false] - will overwrite the target directory if true, by default false
+ * @property {object} [variables = {}] - variables to be resolved in ejs template
  */
 
 const fs = require("fs")
 const path = require("path")
-const glob = require("glob")
+const glob = require("fast-glob")
 const evaluate = require("./eval.js")
 
+class Generator{
+  /**
+   * @param {string} templateDir
+   * @param {string} outDir
+   * @param {Configuration} config
+   */
+  constructor(templateDir, outDir, config){
+    this.templateDir = templateDir
+    this.outDir = outDir
+    this.config = config
+    this.fileMap = {}
+
+    this.init("**/*", outDir)
+  }
+
+  /**
+   * Setup generator
+   *
+   * @param {string} wpath
+   * @param {string} outpath
+   * @returns {Generator}
+   */
+  init(wpath, outpath){
+    const options = {
+      recursive: this.config.overwrite,
+    }
+
+    fs.mkdirSync(this.outDir, options)
+
+    const res = glob.sync(wpath, { cwd: this.templateDir, onlyFiles: true })
+
+    for (const item of res){
+      let dirname = path.dirname(item)
+      let filename = path.basename(item)
+      if (filename[0] === "_"){
+        filename = filename.substr(1)
+      }
+
+      let newPath = path.join(outpath, dirname, filename)
+
+      this.fileMap[item] = newPath
+    }
+
+    return this
+  }
+
+  /**
+   * Walks through wpath and adds files to fileMap
+   *
+   * @param {string} wpath
+   * @param {string} outpath
+   */
+  walk(wpath, outpath){
+    const res = glob.sync(wpath, { cwd: this.templateDir, onlyFiles: true })
+    for (const item of res){
```

Comment:
Can't we have a condition to check if the `outPath` is a path to a file and not a folder before we enter into the loop so that it saves us from the iteration as ultimately the result will be moving the last item of `res` to that `outPath`?

Reply:
:+1: also, `/` and `\` can be replaced by [path.sep](https://nodejs.org/api/path.html#path_path_sep)